### DISCUSSION TITLE
Database Windows

### DISCUSSION THREAD

**Discussion by matty0ung** [Brewtarget/brewtarget#618] [Q&A]
When you click on one of the icons in the top row of the main window, you get a pop-up window that is, sometimes, an editor (eg for Equipments, Mashes and Styles) and sometimes a sortable/filtrable list of all items of this type (eg for Hops/Fermentables/Yeasts/Miscs).  A couple of questions:

- Is it intentional that these are different, or did they just get written at different times?
- For the Hops/Fermentables/Yeasts/Miscs, are some of the columns not meant to be shown?  Eg, I didn't really understand the Time & Amount columns for Misc & Hop
- On Fermentables, we show Method and Addition (sometimes with one greyed out where the other limits it) but when you edit a Fermentable, you get a different way of looking at things (tick boxes for Add After Boil, Recommend Mash and Is Mashed).  Is it intentional that this is different?

**Comment by mikfire** [comment]
It is both historic and intentional. I actually don't think they were written at different times. Most of that was already in place when I joined the project. I will try to summarize this quickly. I may not succeed.

Those windows were the only way to do anything to what we now call a NamedEntity -- modify it, add it to a recipe, etc. -- until I added the trees somewhere in version 2.1. They are designed and optimized for adding a thing to a recipe. Notice that you get exactly the same window for clicking the "+" button next to the fermentables table as you do for clicking the fermentable button in the toolbar. 

The reason the you get two different kinds of interfaces is that the things are presented in two different interfaces. Fermentables, hops, misc and yeasts are all shown as tables in the main window. When you click the "+" button for a fermentable, you get a table with exactly the same columns as we show in the main window. If you want more information, you have to click through. The others are shown in the main window as drop down buttons, so we skip that intermediary window and take you straight to the editor.

Yes. Columns are meant to not be shown in that window. They are intended to look just like the tables in the main window. I would also argue we don't need to know everything about an ingredient when adding it to a recipe -- I probably don't really need to know the diastatic power or the % protein when adding a fermentable to a recipe. I probably do care about the color.

I have no answer for your third question. It's just the way it's been done.

**Reply by matty0ung** [reply]
Gosh, I can't think of the program without the trees down the side.  It's the main way I access stuff!
OK, makes sense for the rest. I agree about not needing huge amounts of info when adding an ingredient.
I'm guessing a lot of fields are things we're going want to come back to when we do BeerJSON, because, AIUI, a full implementation will imply a fair few enhancements to the data model (hence my desire to do the database layer stuff first).  So maybe that will be a good time to have another look at some of these bits of the UI.

**Comment by matty0ung** [comment]
These Catalog / Database windows have been reworked slightly as part of the BeerJSON work, so archiving this discussion now.
